hw/opentitan/ot_usbdev.c

```diff
@@ -429,6 +512,225 @@ static void ot_usbdev_update_alerts(OtUsbdevState *s)
+
+    /* VBUS was turned on */
+    if (vbus_sense) {
+        /* See BFM (bus_connect) */
+        g_assert(ot_usbdev_get_link_state(s) ==
+                 OT_USBDEV_LINK_STATE_DISCONNECTED);
+        s->regs[R_USBDEV_INTR_STATE] |= USBDEV_INTR_POWERED_MASK;
+
+        if (ot_usbdev_is_enabled(s)) {
+            ot_usbdev_set_raw_link_state(s, OT_USBDEV_LINK_STATE_POWERED);
+            ot_usbdev_server_report_connected(s, true);
+        }
+    }
+    /* VBUS was turned off */
+    else {
+        /* See BFM (bus_disconnect) */
+        g_assert(ot_usbdev_get_link_state(s) !=
+                 OT_USBDEV_LINK_STATE_DISCONNECTED);
+        ot_usbdev_set_raw_link_state(s, OT_USBDEV_LINK_STATE_DISCONNECTED);
+        s->regs[R_USBDEV_INTR_STATE] |= USBDEV_INTR_DISCONNECTED_MASK;
+        s->regs[R_USBCTRL] =
+            FIELD_DP32(s->regs[R_USBCTRL], USBCTRL, DEVICE_ADDRESS, 0);
+    }
+
+    ot_usbdev_update_irqs(s);
+}
+
+/*
+ * Update the VBUS gate.
+ *
+ * Changes will trigger events (see ot_usbdev_update_vbus).
+ *
+ * @gate New value of the gate.
+ */
+static void ot_usbdev_set_vbus_gate(OtUsbdevState *s, bool gate)
+{
+    s->vbus_gate = gate;
+
+    ot_usbdev_update_vbus(s);
+}
+
+/*
+ * Simulate a link reset.
+ *
+ * This function will trigger all necessary state and IRQs changes necessary to
+ * perform a link reset.
+ *
+ * @todo document what happens to transfers when done
+ */
+static void ot_usbdev_simulate_link_reset(OtUsbdevState *s)
+{
+    g_assert(!resettable_is_in_reset(OBJECT(s)));
+
+    /* We cannot simulate a reset if the device is disconnected! */
+    if (ot_usbdev_get_link_state(s) == OT_USBDEV_LINK_STATE_DISCONNECTED) {
+        error_report("%s: %s Link reset while disconnected?!", __func__,
+                     s->ot_id);
+        return;
+    }
+
+    s->regs[R_USBDEV_INTR_STATE] |= USBDEV_INTR_LINK_RESET;
+
+    /* See BFM (bus_reset) */
+    s->regs[R_USBCTRL] =
+        FIELD_DP32(s->regs[R_USBCTRL], USBCTRL, DEVICE_ADDRESS, 0);
+    s->regs[R_USBDEV_INTR_STATE] |= USBDEV_INTR_LINK_RESET_MASK;
+    /* @todo cancel all transfers */
+
+    if (ot_usbdev_has_vbus(s) && ot_usbdev_is_enabled(s)) {
+        /* @todo BFM has some extra state processing but it seems incorrect */
+        ot_usbdev_set_raw_link_state(s, OT_USBDEV_LINK_STATE_ACTIVE_NOSOF);
+    }
+    ot_usbdev_update_irqs(s);
+}
+
+/*
+ * Update the value of the USBCTRL register.
+ *
+ * This function will update the USBCTRL register after a write and
+ * trigger the necessary changes if the ENABLE bit is changed. If
+ * as a result of this change the device becomes (dis)connected,
+ * the server will be notified.
+ *
+ * @val32 New value
+ */
+static void ot_usbdev_write_usbctrl(OtUsbdevState *s, uint32_t val32)
+{
+    bool old_enable = ot_usbdev_is_enabled(s);
+
+    /* @todo Handle resume_link_active eventually, this is a W/O field */
+    s->regs[R_USBCTRL] = val32 & USBDEV_USBCTRL_R_MASK;
```

Comment:
This todo is fine, but maybe instead just save some potential future trouble with a little extra logic?

```suggestion
    s->regs[R_USBCTRL] &= ~USBDEV_USBCTRL_R_MASK;
    s->regs[R_USBCTRL |= val32 & USBDEV_USBCTRL_R_MASK;
```

Reply:
Hum, that seems excessively paranoid since the non-read bits will never set in the register (they just trigger stuff on writes)

Reply:
Sorry, this is definitely a reading comprehension failure on my part, I thought this comment said that the resume link was `R/O`, not `W/O`, and so was using this logic to ensure we only set writable bits (not to be excessively paranoid...). Sorry!